gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java

```diff
@@ -190,20 +215,55 @@ public int compare(URI c1, URI c2) {
       return ImmutableList.of();
     }
 
-    List<ConfigBasedDataset> result = new ArrayList<ConfigBasedDataset>();
-    for (URI leaf : leafDatasets) {
-      try {
-        result.add(new ConfigBasedDataset(configClient.getConfig(leaf)));
-      } catch (VersionDoesNotExistException | ConfigStoreFactoryDoesNotExistsException
-          | ConfigStoreCreationException e) {
-        log.error("Caught error while retrieve config for " + leaf + ", skipping.");
-      }
-    }
+    final List<CopyableDatasetBase> result = new CopyOnWriteArrayList<>();
+
+    Iterator<Callable<Void>> callableIterator =
+        Iterators.transform(leafDatasets.iterator(), new Function<URI, Callable<Void>>() {
+          @Override
+          public Callable<Void> apply(final URI datasetURI) {
+            return findDatasetsCallable(configClient, datasetURI, props, result);
+          }
+        });
+
+    this.executeItertorExecutor(callableIterator);
+    log.info(String.format("found %s datasets in ConfigBasedDatasetsFinder ", result.size()));
     return result;
   }
 
   @Override
   public Path commonDatasetRoot() {
     return this.replicationDatasetCommonRoot;
   }
+
+  private void executeItertorExecutor(Iterator<Callable<Void>> callableIterator) throws IOException {
+    try {
+      List<Future<Void>> futures = new IteratorExecutor<>(callableIterator, this.threadPoolSize,
```

Comment:
If you do go this route it is easier to do `IteratorExecutor.logFailures(iteratorExecutor.executeAndGetResults(), logger, 100)`